### ISSUE TITLE
Mutated values in where clause

### ISSUE THREAD

**Issue by silentlight** [vuex-orm/vuex-orm#19]
Hi @kiaking,

Is there any way to use `mutated` value inside of the `where` clause?
I have a datetime attributes for which I use `mutators` to convert them to moment objects. It would be nice to use these moment objects inside my where clause so that I can simply do something like this: `.where(record => record.started_at.isBetween(self.startTime, self.endTime))`. Of course I can convert `record.started_at` to moment object inside of where clause but I figured that maybe mutators will cover this case.

**Comment by silentlight** [comment]
Interestingly, I cannot convert to moment object inside of `where` clause like so:
`.where(record => moment(record.started_at).isBetween(self.startTime, self.endTime))` because I get `"ReferenceError: __WEBPACK_IMPORTED_MODULE_0_moment___default is not defined"` error.

**Comment by CinKon** [comment]
The moment-Library was removed to to unnecessary dependency-load. #8 
You have to mutate the values yourself

**Comment by CinKon** [comment]
Or add moment.js manual

**Comment by silentlight** [comment]
Yes, I added moment.js manually and inside of my model I have:
```
static mutators() {
    return {
      started_at(value) {
        return moment(value);
      },
      stopped_at(value) {
        return moment(value);
      },
    };
  }
```

**Comment by kiaking** [comment]
Hi @silentlight 

First thing first, I think you have to import `moment` where you define your where clause.

```js
// Import moment.
import moment from 'moment'

// Then use it in the where clause.
store.getters['entities/users/query']().where((record) => {
  moment(record.started_at).isBetweeen(self.startTime, self.endTime)
}).get()
```

Also, make sure to upgrade Vuex ORM to 0.11.1 which was released just a few hours ago. 0.11.1 fixed an issue where you couldn't use local variables in where clause.

Next about original issue where you wanted to use mutated value in where clause. Your example using moment sounds very nice, and I too think that's a nice feature to have. However, I want to have a little more discussion about this topic.

The reason why you can't use mutated value in where clause right now is because mutation happens when instantiating the models, which happens **after** where clause has resolved.

We can make it able to use mutated values in where clause by instantiating models before it's passed to where clause but this requires creating whole lot of class instances.

Also, mutating date with `moment` sounds very nice, but I think maybe there is a time user don't want to have value mutated.

For example, how about this case?

```js
class User extends Model {
  static entity = 'users'

  static fields () {
    return {
      id: this.attr(null),
      role: this.attr(0)
    }
  }

  static mutators () {
    return {
      role (value) {
        return value === 0 ? 'sysadmin' : 'admin'
      }
    }
  }
}
```

In this case, maybe there's a chance where a user wants to compare the value of `role` by Number instead fo String?

```js
store.getters['entities/users/query']()
  // Should compare with string or number?
  .where('role', 1) // or .where('role', 'sysadmin') ?
```

If we make value passed to where clause mutated, a user will lose a way to compare against the original value in where clause.

So I think it's better to leave the value in where clause as is for now since you can always mutate the value however you want in where clause.

What do you think?

**Comment by silentlight** [comment]
@kiaking My problem has been solved when I updated to 0.11.1 version. Now I can use `moment` inside where clauses. That's great! Thanks!

Going back to using mutated values inside of where clause, I think you brought up good point. If we only use mutated values than there will be no way to get an original value in where clause. 
We could also add second parameter to where closure like so:

```
store.getters['entities/users/query']().where((record, originalRecord) => {
  return record.isBetweeen(self.startTime, self.endTime)
}).get()
```

I am not sure about this second parameter, but I just wanted to share this idea with you.

**Comment by kiaking** [comment]
> My problem has been solved when I updated to 0.11.1 version. Now I can use moment inside where clauses. That's great! Thanks!

Nice! Great to hear that : )

> We could also add second parameter to where closure like so:

Hah, yeah that would solve the problem. Nice idea!

I would instead make the first argument to be the original value since if there's no mutator, two arguments are going to be exactly same.

The last thing to consider is that we are going to have to instantiate a new class every time there's closure in where clause. I guess mainly this is going to impact performance when there're lots of records (if you have 100 records to fetch, you have to instantiate 100 classes).

That's being said; I would like to implement this feature. I don't think people are going to store thousands of records to the vuex anyway. If the performance becomes an issue, we can think of some another workaround at that time (maybe there is a way to check if 2nd arguments exist, and if not we can skip instantiating the class for that closure).

**Comment by silentlight** [comment]
@kiaking Sounds good. I think that your idea about checking second argument could do the thing :)

**Comment by kiaking** [comment]
It has been a while, but this feature is now available with Vuex ORM 0.16.2!
https://github.com/vuex-orm/vuex-orm/releases/tag/v0.16.2
